File: shop-manager/backend/app/core/envelope.py
"""Universal Response Envelope (Standards: CODING_STANDARDS §4.2).

Every API response — success or failure — uses the same shape so clients
parse one contract:

    {"success": true,  "data": <payload>, "error": null}
    {"success": false, "data": null,      "error": {"code", "message", "details"}}
"""
from typing import Any


def success(data: Any = None) -> dict:
    return {"success": True, "data": data, "error": None}


def failure(code: str, message: str, details: Any = None) -> dict:
    return {
        "success": False,
        "data": None,
        "error": {"code": code, "message": message, "details": details},
    }

File: mobile/admin-app/app/src/main/java/com/arvshop/admin/data/remote/Envelope.java
package com.arvshop.admin.data.remote;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Parses V01's universal response envelope:
 *   {"success":true,  "data":<T>, "error":null}
 *   {"success":false, "data":null, "error":{"code","message","details"}}
 * Pure logic — unit-tested in EnvelopeTest against fixtures captured from the live API.
 */
public final class Envelope {

    private Envelope() { }

    /**
     * @return the "data" object on success.
     * @throws ApiException carrying the backend error code + http status on failure.
     */
    public static Object unwrap(String body, int httpStatus) throws ApiException {
        JSONObject root;
        try {
            root = new JSONObject(body);
        } catch (JSONException e) {
            throw new ApiException(ApiException.CODE_PARSE,
                    "Unexpected server response", httpStatus);
        }
        boolean success = root.optBoolean("success", false);
        if (success && httpStatus >= 200 && httpStatus < 300) {
            return root.opt("data");
        }
        JSONObject error = root.optJSONObject("error");
        String code = error != null ? error.optString("code", "ERROR") : "ERROR";
        String message = error != null ? error.optString("message", "Request failed")
                : "Request failed (HTTP " + httpStatus + ")";
        throw new ApiException(code, message, httpStatus);
    }

    public static JSONObject dataObject(String body, int httpStatus) throws ApiException {
        Object data = unwrap(body, httpStatus);
        if (!(data instanceof JSONObject)) {
            throw new ApiException(ApiException.CODE_PARSE, "Expected an object", httpStatus);
        }
        return (JSONObject) data;
    }
}
